Xamarin.Forms.Platform.Android/CollectionView/EmptyViewAdapter.cs

```diff
@@ -62,50 +112,162 @@ public override void OnViewRecycled(Object holder)
+		{
+			if (IsHeader(position))
+			{
+				return _headerViewType;
+			}
+
+			if (IsFooter(position))
+			{
+				return _footerViewType;
+			}
+
+			if (IsEmpty(position))
+			{
+				return _emptyItemViewType;
+			}
+
+			return base.GetItemViewType(position);
+		}
+
+		protected RecyclerView.ViewHolder CreateHeaderFooterViewHolder(object content, DataTemplate template, Context context)
+		{
+			if (template != null)
+			{
+				var footerContentView = new ItemContentView(context);
+				return new TemplatedItemViewHolder(footerContentView, template, isSelectionEnabled: false);
+			}
+
+			if (content is View formsView)
+			{
+				return SimpleViewHolder.FromFormsView(formsView, context);
+			}
+
+			// No template, Footer is not a Forms View, so just display Footer.ToString
+			return SimpleViewHolder.FromText(content?.ToString(), context, false);
+		}
+
+		protected RecyclerView.ViewHolder CreateEmptyViewHolder(object content, DataTemplate template, ViewGroup parent)
+		{
+			var context = parent.Context;
 
 			if (template == null)
 			{
-				if (!(EmptyView is View formsView))
+				if (!(content is View formsView))
 				{
 					// No template, EmptyView is not a Forms View, so just display EmptyView.ToString
-					return SimpleViewHolder.FromText(EmptyView?.ToString(), context);
+					return SimpleViewHolder.FromText(content?.ToString(), context, () => parent.Width, () => parent.Height - GetHeaderFooterHeight(parent));
 				}
 
 				// EmptyView is a Forms View; display that
-				return SimpleViewHolder.FromFormsView(formsView, context, () => parent.Width, () => parent.Height);
+				return SimpleViewHolder.FromFormsView(formsView, context, () => parent.Width, () => parent.Height - GetHeaderFooterHeight(parent));
 			}
 
-			var itemContentView = new SizedItemContentView(parent.Context, () => parent.Width, () => parent.Height);
+			var itemContentView = new SizedItemContentView(parent.Context, () => parent.Width, () => parent.Height - GetHeaderFooterHeight(parent));
 			return new TemplatedItemViewHolder(itemContentView, template, isSelectionEnabled: false);
 		}
 
-		public override int GetItemViewType(int position)
+		protected void BindTemplatedItemViewHolder(TemplatedItemViewHolder templatedItemViewHolder, object context)
+		{
+			templatedItemViewHolder.Bind(context, ItemsView);
+		}
+
+		bool IsHeader(int position)
+		{
+			if (Header == null && HeaderTemplate == null)
+				return false;
+
+			return position == 0;
+		}
+
+		bool IsFooter(int position)
 		{
-			return _itemViewType;
+			if (Footer == null && FooterTemplate == null)
+				return false;
+
+			return position == ItemCount - 1;
+		}
+
+		bool IsEmpty(int position)
+		{
+			if (EmptyView == null && EmptyViewTemplate == null)
+				return false;
+
+			return (Header == null && HeaderTemplate == null) ? position == 0 : position == 1;
+		}
+
+		int GetHeaderFooterHeight(ViewGroup parent)
+		{
+			int height = 0;
+
+			if (parent.ChildCount > 0)
```

Comment:
It looks like this only gets the height of the header; shouldn't it also include the footer height (if there _is_ a footer)? And shouldn't it account for the footer if there is no header?

Reply:
Changed, now it takes into account header and footer height.